pkg/portal/middleware/aad.go

```diff
@@ -286,35 +221,63 @@ func (a *aad) callback(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	idToken, err := a.verifier.Verify(r.Context(), rawIDToken)
+	_, _, _, expiry, err := a.accessValidator.validateAccess(r.Context(), rawIDToken, a.now())
 	if err != nil {
-		a.internalServerError(w, err)
+		// we could not identify the user so we make them try to login again
+		http.Redirect(w, r, "/api/login", http.StatusTemporaryRedirect)
 		return
 	}
 
+	// only keep the cookie for the validity time of the token
+	http.SetCookie(w, &http.Cookie{
+		Name:     OIDCCookie,
+		Value:    rawIDToken,
+		Expires:  expiry,
+		SameSite: http.SameSiteStrictMode,
+		Secure:   true,
+		HttpOnly: true,
+	})
+
+	// delete the state cookie
+	http.SetCookie(w, &http.Cookie{
+		Name:   stateCookie,
+		Value:  "",
+		MaxAge: -1,
+	})
+
+	http.Redirect(w, r, "/", http.StatusTemporaryRedirect)
+}
+
+func extractInfoFromToken(token *oidc.IDToken) (preferredUsername string, groups []string, err error) {
 	var claims claims
-	err = idToken.Claims(&claims)
-	if err != nil {
-		a.internalServerError(w, err)
-		return
-	}
 
-	groupsIntersect := GroupsIntersect(a.allGroups, claims.Groups)
-	if len(groupsIntersect) == 0 {
-		http.Error(w, http.StatusText(http.StatusForbidden), http.StatusForbidden)
+	err = token.Claims(&claims)
+	if err != nil {
+		return "", nil, err
 	}
 
-	session.Values[SessionKeyUsername] = claims.PreferredUsername
-	session.Values[SessionKeyGroups] = groupsIntersect
-	session.Values[SessionKeyExpires] = a.now().Add(a.sessionTimeout).Unix()
+	return claims.PreferredUsername, claims.Groups, nil
```

Comment:
why not return the entire claim object?

Reply:
it just avoids having to deal with a claim later on, because after that we really only need about the username and the groups, not the claim.